core/domain/prod_validation_jobs_one_off.py

```diff
@@ -15,28 +15,133 @@
+            models.NAMES.recommendations, models.NAMES.story,
+            models.NAMES.user]))
 datastore_services = models.Registry.import_datastore_services()
 
+ALLOWED_AUDIO_EXTENSIONS = feconf.ACCEPTED_AUDIO_EXTENSIONS.keys()
+ALLOWED_IMAGE_EXTENSIONS = list(itertools.chain.from_iterable(
+    feconf.ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS.values()))
+ASSETS_PATH_REGEX = '/exploration/[A-Za-z0-9]{1,12}/assets/'
+IMAGE_PATH_REGEX = (
+    '%simage/[A-Za-z0-9_]{1,}\\.(%s)' % (
+        ASSETS_PATH_REGEX, ('|').join(ALLOWED_IMAGE_EXTENSIONS)))
+AUDIO_PATH_REGEX = (
+    '%saudio/[A-Za-z0-9_]{1,}\\.(%s)' % (
+        ASSETS_PATH_REGEX, ('|').join(ALLOWED_AUDIO_EXTENSIONS)))
+FILE_MODELS_REGEX = '(%s|%s)' % (IMAGE_PATH_REGEX, AUDIO_PATH_REGEX)
+
 
 class BaseModelValidator(object):
     """Base class for validating models."""
 
-    errors = {}
-    # external_models is keyed by field name. Each value consists
-    # of the model class and a list of (external_key, external_model_instance)
+    errors = collections.defaultdict(list)
+    # external_instance_details is keyed by field name. Each value consists
+    # of a list of (model class, external_key, external_model_instance)
     # tuples.
-    external_models = {}
+    external_instance_details = {}
+
+    @classmethod
+    def _get_model_id_regex(cls, unused_item):
+        """Returns a regex for model id.
+
+        This method can be overridden by subclasses, if needed.
+
+        Args:
+            unused_item: ndb.Model. Entity to validate.
+
+        Returns:
+            str. A regex pattern to be followed by the model id.
+        """
+        return '^[A-Za-z0-9]{1,%s}$' % base_models.ID_LENGTH
+
+    @classmethod
+    def _validate_model_id(cls, item):
+        """Checks whether the id of model matches the regex specified for
+        the model.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+        """
+        regex_string = cls._get_model_id_regex(item)
+        if not re.compile(regex_string).match(str(item.id)):
+            cls.errors['model id check'].append((
+                'Entity id %s: Entity id does not match regex pattern') % (
+                    item.id))
+
+    @classmethod
+    def _get_model_domain_object_instance(cls, unused_item):
+        """Returns a domain object instance created from the model.
+
+        This method can be overridden by subclasses, if needed.
+
+        Args:
+            unused_item: ndb.Model. Entity to validate.
+
+        Returns:
+            *: A domain object to validate.
+        """
+        return None
+
+    @classmethod
+    def _validate_model_domain_object_instances(cls, item):
+        """Checks that model instance passes the validation of the domain
+        object for model.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+        """
+        model_domain_object_instance = cls._get_model_domain_object_instance(
+            item)
+
+        if not model_domain_object_instance:
```

Comment:
if model_domain_object_instance is None

And then, before the return, put a comment below saying "No domain object exists for this storage model class."

Reply:
Done.